keps/sig-cli/0031-datadrivencommands.md

```diff
@@ -0,0 +1,412 @@
+- Are Agnostic to Specific Resources
+
+## Proposal
+
+Define a format for publishing simple cli commands as data.  CLI commands would be limited to:
+
+- Sending one or more requests to Resource or SubResource Endpoints
+- Populating requests from command line flags and arguments
+- Writing output populated from the Responses
+
+**Proof of Concept:** [cnctl](https://github.com/pwittrock/kubectl/tree/cnctl/cmd/cnctl)
+
+Instructions to run PoC:
+
+- `go run ./main.go` (no commands show up)
+- `kubectl apply` the `cli_v1alpha1_clitestresource.yaml` (add the CRD with the commands)
+- `go run ./main.go` (create command shows up)
+- `go run ./main create deployment -h` (view create command help)
+- `go run ./main create deploy --image nginx --name nginx` (create a deployment)
+- `kubectl get deployments`
+
+### Implementation Details
+
+**Publishing Data:**
+
+Alpha: No apimachinery changes required
+
+- Alpha: publish extension Resource Commands as an annotation on CRDs.
+- Alpha: publish core Resource Commands as openapi extension.
+
+Beta: apimachinery changes required
+
+- Beta: publish extension Resource Commands a part of the CRD Spec.
+- Beta: publish core Resource Commands from new endpoint (like *swagger.json*)
+
+**Data Command Structure:**
+
+```go
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// ResourceCommand defines a command that is dynamically defined as an annotation on a CRD
+type ResourceCommand struct {
+	// Command is the cli Command
+	Command Command `json:"command"`
+
+	// Requests are the requests the command will send to the apiserver.
+	// +optional
+	Requests []ResourceRequest `json:"requests,omitempty"`
+
+	// Output is a go-template used write the command output.  It may reference values specified as flags using
+	// {{index .Flags.Strings "flag-name"}}, {{index .Flags.Ints "flag-name"}}, {{index .Flags.Bools "flag-name"}},
+	// {{index .Flags.Floats "flag-name"}}.
+	//
+	// It may also reference values from the responses that were saved using saveResponseValues
+	// - {{index .Responses.Strings "response-value-name"}}.
+	//
+	// Example:
+	// 		deployment.apps/{{index .Responses.Strings "responsename"}} created
+	//
+	// +optional
+	Output string `json:"output,omitempty"`
+}
+
+type ResourceOperation string
+
+const (
+	CREATE_RESOURCE ResourceOperation = "Create"
+	UPDATE_RESOURCE                   = "Update"
+	DELETE_RESOURCE                   = "Delete"
```

Comment:
Boo, hiss.  Use CamelCase, we’re not C savages here.

Reply:
lolz, will update

Reply:
Done